helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterContext.java

```diff
@@ -19,9 +19,104 @@
+  // This estimation helps to ensure global top state replica count evenness
+  private final int _estimatedMaxTopStateCount;
+  // This estimation helps to ensure per-resource partition count evenness
+  private final Map<String, Integer> _estimatedMaxPartitionByResource = new HashMap<>();
+
+  // map{zoneName : map{resourceName : set(partitionNames)}}
+  private Map<String, Map<String, Set<String>>> _assignmentForFaultZoneMap = new HashMap<>();
+
+  /**
+   * Construct the cluster context based on the current instance status.
+   *
+   * @param replicaSet    All the partition replicas that are managed by the rebalancer
+   * @param instanceCount The count of all the active instances that can be used to host partitions.
+   */
+  ClusterContext(Set<AssignableReplica> replicaSet, int instanceCount) {
+    int totalReplicas = 0;
+    int totalTopStateReplicas = 0;
+
+    for (Map.Entry<String, List<AssignableReplica>> entry : replicaSet.stream()
+        .collect(Collectors.groupingBy(AssignableReplica::getResourceName)).entrySet()) {
+      int replicas = entry.getValue().size();
+      totalReplicas += replicas;
+
+      int replicaCnt = Math.max(1, estimateAvgReplicaCount(replicas, instanceCount));
+      _estimatedMaxPartitionByResource.put(entry.getKey(), replicaCnt);
+
+      totalTopStateReplicas +=
+          entry.getValue().stream().filter(AssignableReplica::isReplicaTopState).count();
+    }
+
+    _estimatedMaxPartitionCount = estimateAvgReplicaCount(totalReplicas, instanceCount);
+    _estimatedMaxTopStateCount = estimateAvgReplicaCount(totalTopStateReplicas, instanceCount);
+  }
+
+  public Map<String, Map<String, Set<String>>> getAssignmentForFaultZoneMap() {
+    return _assignmentForFaultZoneMap;
+  }
+
+  public Integer getEstimatedMaxPartitionCount() {
+    return _estimatedMaxPartitionCount;
+  }
+
+  public Integer getEstimatedMaxPartitionByResource(String resourceName) {
+    return _estimatedMaxPartitionByResource.get(resourceName);
+  }
+
+  public Integer getEstimatedMaxTopStateCount() {
+    return _estimatedMaxTopStateCount;
+  }
+
+  public Set<String> getPartitionsForResourceAndFaultZone(String faultZoneId, String resourceName) {
+    if (_assignmentForFaultZoneMap.containsKey(faultZoneId) && _assignmentForFaultZoneMap
+        .get(faultZoneId).containsKey(resourceName)) {
+      return _assignmentForFaultZoneMap.get(faultZoneId).get(resourceName);
+    }
+    return Collections.emptySet();
+  }
+
+  void addPartitionToFaultZone(String faultZoneId, String resourceName, String partition) {
+    if (!_assignmentForFaultZoneMap.containsKey(faultZoneId)) {
+      _assignmentForFaultZoneMap.put(faultZoneId, new HashMap<>());
+    }
+    if (!_assignmentForFaultZoneMap.get(faultZoneId).containsKey(resourceName)) {
+      _assignmentForFaultZoneMap.get(faultZoneId).put(resourceName, new HashSet<>());
+    }
+    _assignmentForFaultZoneMap.get(faultZoneId).get(resourceName).add(partition);
```

Comment:
For this code block, would it look cleaner if we `computeIfAbsent()`?
```java
_assignmentForFaultZoneMap.computeIfAbsent(faultZoneId, k -> new HashMap<>())
                          .computeIfAbsent(resourceName, k -> new HashSet<>())
                          .add(partition);
```

Reply:
Looks great. I will use it more.